Do not fuzz ASTDictionaryLayout

akuzm opened the pull request:
+ some cosmetic changes

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

akuzm committed:
Do not fuzz ASTDictionaryLayout

+ some cosmetic changes

alexey-milovidov commented on programs/client/QueryFuzzer.cpp:
```diff
@@ -305,6 +306,26 @@ void QueryFuzzer::fuzz(ASTPtr & ast)
+        // Don't fuzz it. It is a somewhat unconventional AST which does not
+        // directly correspond to the query text, but performs some validation
+        // by itself. It considers its children to be ASTLiteral * only, so when
+        // we replace them with something column-like, the client segfaults.
```
But it should not crash, why don't just throw exception about wrong AST?

akuzm replied:
That is the problem -- the AST is validated at the parsing stage, so if you change it after that, you get a segfault. It is not possible to trigger from SQL. Normally we keep the AST close to what is written, and validate it later -- this is the only case I encountered so far that doesn't work like this.

alexey-milovidov replied:
Why don't add yet another validation even if it is not needed at all in normal runs?

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
fixup

akuzm committed:
wip

akuzm committed:
fixes

akuzm committed:
Merge remote-tracking branch 'origin/master' into HEAD

akuzm committed:
fixes

akuzm commented:
* ast fuzzer failure is https://github.com/ClickHouse/ClickHouse/issues/13248
* perf test is broken in master
* 00933_test_fix_extra_seek_on_compressed_cache is in CI Fest #7 https://github.com/ClickHouse/ClickHouse/issues/12508

akuzm commented:
So what I did eventually is:

* Removed a potentially unsafe transformation from query fuzzer (replacing random child ASTLiteral with another AST node). See the comment on why it is not safe.
* Fixed some problems in dictionary ASTs including broken clone() methods.

I'll merge this now.
